### ISSUE TITLE
Some video can't be downloaded.

### ISSUE THREAD

**Issue by tridorje** [Cyklet/VidLoader#29]
Please help me to check case 1 that can't be download video.
Case 1: The below video can't be downloaded

> `Parameter of error data
> Printing description of downloadValue:
> ▿ DownloadValues
>   - identifier : "216489"
>   ▿ url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216489-923441&p=6141&d=629&n=1698&t=4a77da6d23a2d344ad0a3b32c9ae4e88&previewflag=0&logging=1
>     - _url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216489-923441&p=6141&d=629&n=1698&t=4a77da6d23a2d344ad0a3b32c9ae4e88&previewflag=0&logging=1
>   - title : "大株主ファンドが応募意向 東京ドームへのＴＯＢ"
>   - artworkData : nil
>   ▿ minRequiredBitrate : Optional<Int>
>     - some : 4096000
> (lldb) 
> 

====>Error Log:

> 2020-12-10 04:40:41137 [Debug] [BODNetworkService.swift] : 45 upload(_:) -> --------------------------------BEGIN------------------------------------------
> 2020-12-10 04:40:41138 [Debug] [BODNetworkService.swift] : 46 upload(_:) -> Optional(http://stg2-txbiz.tv-tokyo.co.jp/apiv3/getDownloadUrl.php)
> 2020-12-10 04:40:41344 [Debug] [BODNetworkService.swift] : 65 upload(_:) -> ------------------------------RESPONSE---------------------------------------
> 2020-12-10 04:40:41344 [Debug] [BODNetworkService.swift] : 66 upload(_:) -> <GETDOWNLOADURL><RESPONSE_CODE>1</RESPONSE_CODE><CACHE_EFFECTIVE_TIME>0</CACHE_EFFECTIVE_TIME><TOTAL_FETCH_COUNT>1</TOTAL_FETCH_COUNT><REQUEST_FETCH_COUNT>1</REQUEST_FETCH_COUNT><DOWNLOAD_URL>https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216489-923441&amp;p=6141&amp;d=629&amp;n=1698&amp;t=840b4617e2a1b0aac8714a647611efdc&amp;previewflag=0&amp;logging=1</DOWNLOAD_URL><DOWNLOAD_PLAY_LIMIT>7</DOWNLOAD_PLAY_LIMIT></GETDOWNLOADURL>
> 2020-12-10 04:40:41345 [Debug] [BODNetworkService.swift] : 67 upload(_:) -> --------------------------------END------------------------------------------
> 2020-12-10 04:40:41350 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:41353 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:41353 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 04:40:41466 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:41467 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 16:40:41.476944+0900 [DEBUG]テレ東BOD[15198:6296767] [tcp] tcp_input [C96.1:3] flags=[R] seq=1709103288, ack=0, win=0 state=LAST_ACK rcv_nxt=1709103288, snd_una=1251447016
> 2020-12-10 16:40:41.477899+0900 [DEBUG]テレ東BOD[15198:6296767] [tcp] tcp_input [C96.1:3] flags=[R.] seq=1709103288, ack=1251447048, win=0 state=CLOSED rcv_nxt=1709103288, snd_una=1251447016
> 2020-12-10 04:40:41499 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:41500 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 16:40:41.669808+0900 [DEBUG]テレ東BOD[15198:6297222] [tcp] tcp_input [C98.1:3] flags=[R] seq=4208072893, ack=0, win=0 state=LAST_ACK rcv_nxt=4208072893, snd_una=2357595353
> 2020-12-10 16:40:41.669932+0900 [DEBUG]テレ東BOD[15198:6297222] [tcp] tcp_input [C98.1:3] flags=[R.] seq=4208072893, ack=2357595385, win=0 state=CLOSED rcv_nxt=4208072893, snd_una=2357595353
> 2020-12-10 16:40:41.693719+0900 [DEBUG]テレ東BOD[15198:6297072] [tcp] tcp_input [C99.1:3] flags=[R] seq=185130319, ack=0, win=0 state=LAST_ACK rcv_nxt=185130319, snd_una=1956460810
> 2020-12-10 16:40:41.696248+0900 [DEBUG]テレ東BOD[15198:6297222] [tcp] tcp_input [C99.1:3] flags=[R.] seq=185130319, ack=1956460842, win=0 state=CLOSED rcv_nxt=185130319, snd_una=1956460810
> 2020-12-10 04:40:41704 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:41704 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 04:40:42618 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216489 at: Optional(2020-12-10 07:40:41 +0000)
> 2020-12-10 04:40:42619 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 04:40:42620 [Debug] [DownloadMyListShareCell.swift] : 68 configureDownloadButton(state:episode:) -> 
> ====>
> error: custom(DownloadEpisodeManager.VidLoaderError(code: 0, description: "DownloadEpisodeManager.ResourceLoadingError.m3u8(DownloadEpisodeManager.M3U8Error.dataConversion)", localizedDescription: "The operation couldn’t be completed. (DownloadEpisodeManager.ResourceLoadingError error 0.)", domain: "DownloadEpisodeManager.ResourceLoadingError"))
> <====
> 2020-12-10 04:40:42622 [Debug] [DownLoadListCell.swift] : 137 configureDownloadButton(state:) -> 
> ====>
> error: custom(DownloadEpisodeManager.VidLoaderError(code: 0, description: "DownloadEpisodeManager.ResourceLoadingError.m3u8(DownloadEpisodeManager.M3U8Error.dataConversion)", localizedDescription: "The operation couldn’t be completed. (DownloadEpisodeManager.ResourceLoadingError error 0.)", domain: "DownloadEpisodeManager.ResourceLoadingError"))
> <====`

Case 2: The next video is OK

> `Parameter of downloadable data:
> Printing description of downloadValue:
> ▿ DownloadValues
>   - identifier : "216581"
>   ▿ url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216581-923802&p=6141&d=629&n=1698&t=b12218522c2bc6fdba35008ee00b0b43&previewflag=0&logging=1
>     - _url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216581-923802&p=6141&d=629&n=1698&t=b12218522c2bc6fdba35008ee00b0b43&previewflag=0&logging=1
>   - title : "米当局 フェイスブックを提訴へ 一部事業の売却求める方針"
>   - artworkData : nil
>   ▿ minRequiredBitrate : Optional<Int>
>     - some : 4096000
> (lldb) `

Log of OK data

> 2020-12-10 04:51:55142 [Debug] [BODNetworkService.swift] : 45 upload(_:) -> --------------------------------BEGIN------------------------------------------
> 2020-12-10 04:51:55143 [Debug] [BODNetworkService.swift] : 46 upload(_:) -> Optional(http://stg2-txbiz.tv-tokyo.co.jp/apiv3/getDownloadUrl.php)
> 2020-12-10 04:51:55280 [Debug] [BODNetworkService.swift] : 65 upload(_:) -> ------------------------------RESPONSE---------------------------------------
> 2020-12-10 04:51:55281 [Debug] [BODNetworkService.swift] : 66 upload(_:) -> <GETDOWNLOADURL><RESPONSE_CODE>1</RESPONSE_CODE><CACHE_EFFECTIVE_TIME>0</CACHE_EFFECTIVE_TIME><TOTAL_FETCH_COUNT>1</TOTAL_FETCH_COUNT><REQUEST_FETCH_COUNT>1</REQUEST_FETCH_COUNT><DOWNLOAD_URL>https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216581-923802&amp;p=6141&amp;d=629&amp;n=1698&amp;t=23d341da155b469f4765c18bf10854c3&amp;previewflag=0&amp;logging=1</DOWNLOAD_URL><DOWNLOAD_PLAY_LIMIT>7</DOWNLOAD_PLAY_LIMIT></GETDOWNLOADURL>
> 2020-12-10 04:51:55282 [Debug] [BODNetworkService.swift] : 67 upload(_:) -> --------------------------------END------------------------------------------
> 2020-12-10 04:51:55289 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55293 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55293 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 04:51:55368 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55368 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 16:51:55.374682+0900 [DEBUG]テレ東BOD[15198:6300482] [tcp] tcp_input [C112.1:3] flags=[R] seq=3947647630, ack=0, win=0 state=LAST_ACK rcv_nxt=3947647630, snd_una=2660998770
> 2020-12-10 04:51:55383 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55383 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 16:51:55.477981+0900 [DEBUG]テレ東BOD[15198:6300745] [tcp] tcp_output [C113.1:3] flags=[R.] seq=1995912193, ack=1844248977, win=2047 state=CLOSED rcv_nxt=1844248977, snd_una=1995912162
> 2020-12-10 16:51:55.515240+0900 [DEBUG]テレ東BOD[15198:6300481] [tcp] tcp_input [C114.1:3] flags=[R] seq=2356524569, ack=0, win=0 state=LAST_ACK rcv_nxt=2356524569, snd_una=3478846
> 2020-12-10 04:51:55524 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55524 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.0 
> 2020-12-10 04:51:55869 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:55869 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 0 progress: 0.13333333333333333 
> 2020-12-10 04:51:56539 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:56539 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 2970512 progress: 0.4 
> 2020-12-10 04:51:56684 [Debug] [BODPlayerView.swift] : 323 hideControlPlayerAfterThreeSeconds() -> Ignore hidden while pausing
> 2020-12-10 04:51:57376 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:57377 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 4653408 progress: 0.6666666666666666 
> 2020-12-10 04:51:58048 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:58049 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 6166928 progress: 0.9333333333333333 
> 2020-12-10 04:51:58337 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:58338 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 7984432 progress: 1.0 
> 2020-12-10 04:51:58338 [Debug] [BODPlayVideoView.swift] : 277 download(_:quality:) -> Download completed total time: 3.1966910362243652
> 2020-12-10 04:51:58351 [Debug] [BODRealmManager.swift] : 68 episodeDLInfo(saveDlInfo:intoEpisodeId:) -> Saved episode 216581 at: Optional(2020-12-10 07:51:55 +0000)
> 2020-12-10 04:51:58352 [Debug] [BODPlayVideoView.swift] : 275 download(_:quality:) -> Downloading... 7984432 progress: 1.0 
> 2020-12-10 04:51:58352 [Debug] [BODPlayVideoView.swift] : 277 download(_:quality:) -> Download completed total time: 3.2101199626922607
>

**Comment by tridorje** [comment]
I think that video data is returned but not a UTF8 type, so this video data can't be converted.

![130601301_484529599619685_6414177226727493218_n](https://user-images.githubusercontent.com/30663674/101743917-2cac4380-3b0e-11eb-8122-56dbcab13f96.png)
![131048448_393041481901777_7260440564544277801_n](https://user-images.githubusercontent.com/30663674/101744011-32a22480-3b0e-11eb-9a89-346ed8a8f6be.png)

**Comment by Cyklet** [comment]
Hi @tridorje, yes it seems that data can't be converted to a string that will lead to a fail, nice finding 👍. It's strange that it's falling, .utf8 is quite universal, did you find why it is falling or expected encoding?  
It seems like someone intentionally encoded the data in another format.

**Comment by tridorje** [comment]
I try to log the data by this encode function. Please check.
```
//
//  DataExtensions.swift
//  VidLoader
//
//  Created by Petre on 01.09.19.
//  Copyright © 2019 Petre. All rights reserved.
//

import Foundation

extension Data {
    var string: String? {
        return String(data: self, encoding: .utf8)
    }
    var stringV2: String? {
        let endcodeArr = [
            String.Encoding.ascii,
            String.Encoding.nextstep,
            String.Encoding.japaneseEUC,
            String.Encoding.utf8,
            String.Encoding.isoLatin1,
            String.Encoding.symbol,
            String.Encoding.nonLossyASCII,
            String.Encoding.shiftJIS,
            String.Encoding.isoLatin2,
            String.Encoding.unicode,
            String.Encoding.windowsCP1251,
            String.Encoding.windowsCP1252,
            String.Encoding.windowsCP1253,
            String.Encoding.windowsCP1254,
            String.Encoding.windowsCP1250,
            String.Encoding.iso2022JP,
            String.Encoding.utf16,
            String.Encoding.utf16BigEndian,
            String.Encoding.utf16LittleEndian,
            String.Encoding.utf32,
            String.Encoding.utf32BigEndian,
            String.Encoding.utf32LittleEndian]
        var result1 = ""
        for item in endcodeArr {
            if let result = String(data: self, encoding: item) {
                let str = String(result.prefix(1000))
                print("")
                print("\n\n==============>\nencodetype: \(item.description) raw type: \(item.rawValue) end\n\(str)\n<============\n\n")
                result1 = str
            }
        }
        print("\n\n\n\n\n=>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>end\n\n\n\n\n\n")
        return result1
    }
    var stringV3: String? {
        var tempType = self.stringEncoding
        if tempType == nil {
            tempType = .utf8
        }
        if let string = String(data: self, encoding: tempType!) {
            print(string)
            return string
        }
        return nil
    }
    var stringEncoding: String.Encoding? {
            var nsString: NSString?
            guard case let rawValue = NSString.stringEncoding(for: self, encodingOptions: nil, convertedString: &nsString, usedLossyConversion: nil), rawValue != 0 else { return nil }
            return .init(rawValue: rawValue)
        }
}
```
```
// MARK: - StreamContentRepresentable
    
    func adjust(data: Data, with baseURL: URL, completion: @escaping (Result<Data, M3U8Error>) -> Void) {
        
        guard let response = data.string else {
            
            guard let response2 = data.stringV2 else {
                return completion(.failure(.dataConversion))
            }
            
            return completion(.failure(.dataConversion))
        }
        let response2 = data.stringV2
        let newResponse = replaceRelativeChunks(response: response, with: baseURL)
        replacePaths(response: newResponse, with: baseURL, completion: { result in
            switch result {
            case .success: completion(result)
            case .failure(let error):
                switch error {
                case .custom, .dataConversion, .keyContentWrong: completion(result)
                case .keyURLMissing: completion(.success(newResponse.data!))
                }
            }
        })
    }
```
The txt file below is the log of normal video case, and error video case.
[error-data.txt](https://github.com/Cyklet/VidLoader/files/5672320/error-data.txt)
[OK-data.txt](https://github.com/Cyklet/VidLoader/files/5672321/OK-data.txt)

**Comment by Cyklet** [comment]
Hi @tridorje the download link expired and I can't debug anymore. Also I added the same plist master file to another server and download from there is working, if you are able to post again the link I will investigate it again.

**Comment by tridorje** [comment]
@Cyklet C
Please check an error data.

> Printing description of downloadValue:
> ▿ DownloadValues
>   - identifier : "216489"
>   ▿ url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216489-923441&p=6141&d=629&n=1698&t=de497f67f6918affb1f58a138741404b&previewflag=0&logging=1
>     - _url : https://www2.uliza.jp/IF/iphone/iPhonePlaylist.m3u8?v=216489-923441&p=6141&d=629&n=1698&t=de497f67f6918affb1f58a138741404b&previewflag=0&logging=1
>   - title : "大株主ファンドが応募意向 東京ドームへのＴＯＢ"
>   - artworkData : nil
>   ▿ minRequiredBitrate : Optional<Int>
>     - some : 4096000
> (lldb)

**Comment by tridorje** [comment]
@Cyklet Please accept my skype request.

**Comment by Cyklet** [comment]
Hi @tridorje, I have created a new tag that fixes parsing of \r\n newline -> 1.0.7
I also opened a new issue related to download that fails at the end -> #31
